libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -239,64 +267,143 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
-		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!variable->isConstant() && !variable->immutable())
-				generator.initializeStateVar(*variable);
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
-	}
+string IRGenerator::copyParametersToMemory(ContractDefinition const& _contract)
+{
+	ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	string params = suffixedVariableNameList(
+		"param_" + m_context.functionName(*_contract.constructor()),
+		0,
+		_contract.constructor()->parameters().size()
+	);
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+	Whiskers t(R"X(
+		let programSize := datasize("<object>")
+		let argSize := sub(codesize(), programSize)
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+		let memoryDataOffset := <allocate>(argSize)
+		codecopy(memoryDataOffset, programSize, argSize)
 
-			<assignToParams> <abiDecode>(memoryDataOffset, add(memoryDataOffset, argSize))
+		<assignToParams> <abiDecode>(memoryDataOffset, add(memoryDataOffset, argSize))
+	)X");
+	t("object", creationObjectName(_contract));
+	t("allocate", m_utils.allocationFunction());
+	t("assignToParams", params.empty() ? "" : "let " + params + " := ");
+	t("abiDecode", abiFunctions.tupleDecoder(_contract.constructor()->functionType(false)->parameterTypes(), true));
+
+	m_baseConstructorParameters[_contract.constructor()] = params;
+
+	return t.render();
+}
+
+string IRGenerator::callConstructor(FunctionDefinition const& _constructor)
+{
+	Whiskers t(R"X(
+		<constructorName>(<params>)
+	)X");
+	t("constructorName", m_context.functionName(_constructor));
+	t("params", m_baseConstructorParameters.at(&_constructor));
+
+	return t.render();
+}
+
+string IRGenerator::constructorCode(ContractDefinition const& _contract)
+{
+	// Initialization of state variables in base-to-derived order.
+	solAssert(!_contract.isLibrary(), "Tried to initialize state variables of library.");
+	using boost::adaptors::reverse;
 
-			<constructorName>(<params>)
-		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.functionName(*constructor));
+	Whiskers constructorCode(R"X(
+		<initStateVariables>
```

Comment:
This still initializes *all* state variables up front, right?
If we really want to reproduce the C++ way, then we should instead do this separately for each contract after evaluating the arguments, but before calling the constructor...

Reply:
I am not sure about best way to do this. As when I call constructor itself, I'll not be in scope where I can define state vars. I can define them up-front and init as first thing after evaluation?

Reply:
First of all as rationale for changing this:
```
contract A { uint x; constructor() { x = 42; } function f() public returns(uint256) { return x; } }
contract B is A { uint y = f(); }
```
I'd expect A to be finished at the time ``B`` initializes state variables and get ``y == 42``, not ``y == 0``.

Regarding scoping, I'd need to check - but it might not do any harm to register them up front.

Reply:
@chriseth What do you think about this?

Reply:
I think it is just about right order of calls in constructor function, this case you shared would work with current one as constructor body of A is executed before constructor body of B

Reply:
For reference: we discussed in a call that we still want to change this, so that the call to ``f()`` in ``B`` is done only *after* A is finalized.